File: designtictactoe.js
//Objective is to design a tic-tac-toe game.


//Design of a tic-tac-toe game using rows, columns, and diagonals.

class TicTacToe {
    //Initialize w/ rows, columns, and 2 diagonals
    constructor(n) {
        this.rows = new Array(n).fill(0)
        this.cols = new Array(n).fill(0)
        this.diagonals = new Array(2).fill(0)
        this.size = n
    }

    //Find the right player and update the tile as well as the 
    //appropriate diagonal if needed
    move(row, col, player) {
        let update = (player == 1) ? 1 : -1
        this.rows[row] += update 
        this.cols[col] += update 
        if (row == col) {
            this.diagonals[0] += update 
        }
        if (row == this.size - col - 1) {
            this.diagonals[1] += update
        }

        if (Math.abs(this.rows[row]) == this.size || Math.abs(this.cols[col]) == this.size || Math.abs(this.diagonals[0]) == this.size || Math.abs(this.diagonals[1]) == this.size) {
            return player
        }

        return 0
    }
}

let map = new TicTacToe()
map.move(0,0,1)
map.move(0,2,2)
map.move(2,2,1)
map.move(1,1,2)
map.move(2,0,1)
map.move(1,0,2)
map.move(2,1,1)

File: designtictactoe.java
//Java Solution

class TicTacToe {

    private int[] rows;
    private int[] cols;
    private int[] diags;
    private int size;
    /** Initialize your data structure here. */
    public TicTacToe(int n) {
        rows = new int[n];
        cols = new int[n];
        diags = new int[2];
        size = n;
    }
    
    /** Player {player} makes a move at ({row}, {col}).
        @param row The row of the board.
        @param col The column of the board.
        @param player The player, can be either 1 or 2.
        @return The current winning condition, can be either:
                0: No one wins.
                1: Player 1 wins.
                2: Player 2 wins. */
    public int move(int row, int col, int player) {
        int update = (player == 1) ? 1 : -1;
        rows[row] += update;
        cols[col] += update;
        if (row == col) {
            diags[0] += update;
        }
        if (row == size - col - 1) {
            diags[1] += update;
        }
        
        if (Math.abs(rows[row]) == size || Math.abs(cols[col]) == size || Math.abs(diags[0]) == size || Math.abs(diags[1]) == size) {
            return player;
        }
        
        return 0;
    }
}
